### PR TITLE
Implement Multithreading

### LINKED ISSUE #87: Multithreaded execution
The scheduler should be able to run every ready task in its own thread there is only one security requierement outside of the scheduler that need to get implemented for this. The TaskContainer needs to hold a mutex that gets locked with each access to the input queue.

### LINKED ISSUE #108: Implement a theadsafe queue
For multithreading we need a threadsafe queue to use instead of the default queue.

### PR DESCRIPTION
**Pull request by Highlife1911**
Closes #87 
Closes #108

### CODE REVIEW

File: source/yats/include/yats/queue.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+
+template <typename ValueType>
+class queue
+{
+    using lock = std::lock_guard<std::mutex>;
+public:
+    using value_type = ValueType;
+
+    ValueType extract()
```
Comment (mflueggen) [inline source/yats/include/yats/queue.h]: Why not rename extract to pop?

File: source/yats/include/yats/scheduler.h [hunk 2]
```diff
@@ -35,20 +37,54 @@ class scheduler
...
+                    task->run();
+                    check_runnable();
+                });
+            }
+        };
+
+        check_runnable();
+        while (true)
+        {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (mflueggen) [inline source/yats/include/yats/scheduler.h]: Why is this line required? Looks like a quick fix,

File: source/yats/include/yats/queue.h (hunk 1 above)
Comment (Highlife1911) [reply source/yats/include/yats/queue.h]: pop has a different interface

File: source/yats/include/yats/scheduler.h (hunk 2 above)
Comment (Highlife1911) [reply source/yats/include/yats/scheduler.h]: It is not perfect but the easiest solution until someone implements #72

**Comment by Highlife1911** [comment]
Renaming the class might be a good idea.

File: source/yats/include/yats/scheduler.h (hunk 2 above)
Comment (mflueggen) [reply source/yats/include/yats/scheduler.h]: Ok. Sry. To just get it running it is totally fine.

File: source/yats/include/yats/queue.h (hunk 1 above)
Comment (mflueggen) [reply source/yats/include/yats/queue.h]: That is true... Then I don't like the C++ interface. For me, pop should return and remove the element. But ok.

**Comment by Highlife1911** [comment]
done
